crates/hyperdrive-math/src/lp/math.rs

```diff
@@ -281,6 +301,652 @@ impl State {
+                Ok(present_value) => present_value,
+                // NOTE: Return zero to indicate that the withdrawal shares redeemed
+                // couldn't be calculated.
+                // Errors are safe from this calculation, panics are not.
+                Err(_) => return Ok(fixed!(0)),
+            };
+
+        // If the ending present value is greater than or equal to the starting
+        // present value, we short-circuit to avoid distributing excess idle.
+        // This edge-case can occur when the share reserves is very close to the
+        // minimum share reserves with a large value of k.
+        if ending_present_value >= starting_present_value {
+            return Ok(fixed!(0));
+        }
+
+        // NOTE: This subtraction is safe since the ending present value is less
+        // than the starting present value and the rhs is rounded down.
+        //
+        // Calculate the amount of withdrawal shares that can be redeemed.
+        let lp_total_supply = active_lp_total_supply + withdrawal_shares_total_supply;
+        Ok(lp_total_supply
+            - lp_total_supply.mul_div_down(ending_present_value, starting_present_value))
+    }
+
+    /// Calculates the share proceeds to distribute to the withdrawal pool
+    /// assuming that all of the outstanding withdrawal shares will be
+    /// redeemed. The share proceeds are calculated such that the LP share
+    /// price is conserved. When we need to round, we round down to err on
+    /// the side of slightly too few shares being paid out.
+    fn calculate_distribute_excess_idle_share_proceeds(
+        &self,
+        current_block_timestamp: U256,
+        active_lp_total_supply: FixedPoint, // just the number of lp tokens
+        withdrawal_shares_total_supply: FixedPoint,
+        max_share_reserves_delta: FixedPoint,
+        max_iterations: u64,
+    ) -> Result<FixedPoint> {
+        // Calculate the LP total supply.
+        let lp_total_supply = active_lp_total_supply + withdrawal_shares_total_supply;
+
+        // NOTE: Round the initial guess down to avoid overshooting.
+        //
+        // We make an initial guess for Newton's method by assuming that the
+        // ratio of the share reserves delta to the withdrawal shares
+        // outstanding is equal to the LP share price. In reality, the
+        // withdrawal pool should receive more than this, but it's a good
+        // starting point. The calculation is:
+        //
+        // x_0 = w * (PV(0) / l)
+        let starting_present_value = self.calculate_present_value(current_block_timestamp)?;
+        let mut share_proceeds =
+            withdrawal_shares_total_supply.mul_div_down(starting_present_value, lp_total_supply);
+
+        // If the pool is net neutral, the initial guess is equal to the final
+        // result.
+        let net_curve_trade =
+            self.calculate_net_curve_trade_from_timestamp(current_block_timestamp)?;
```

Comment:
Do we need to use the `safe` variant?

Reply:
So this actually was the safe version, it was just mislabelled.  I updated the name and also added in the unsafe version which `distribute_excess_idle_withdrawal_shares_redeemed` uses.